test/new-e2e/utils/e2e/e2e.go

```diff
@@ -0,0 +1,193 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package e2e provides tools to manage environments and run E2E tests.
+// See [Suite] for an example of the usage.
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/credentials"
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/e2e/client"
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/infra"
+	"github.com/DataDog/test-infra-definitions/aws"
+	"github.com/DataDog/test-infra-definitions/common/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+// Suite manages the environment creation and runs E2E tests.
+// It is implemented as a [testify Suite].
+// Example of usage:
+//
+//	  type MyEnv struct {
+//		   VM *client.VM
+//	  }
+//	  type vmSuite struct {
+//		   *Suite[MyEnv]
+//	  }
+//
+//	  func TestE2ESuite(t *testing.T) {
+//		   suite.Run(t, &vmSuite{Suite: NewSuite("my-test", &StackDefinition[MyEnv]{
+//		     EnvCloudName: "aws/sandbox",
+//			 EnvFactory: func(ctx *pulumi.Context) (*MyEnv, error) {
+//				vm, err := ec2vm.NewUnixLikeEc2VM(ctx, ec2vm.WithOS(os.AmazonLinuxOS, commonos.AMD64Arch))
+//				if err != nil {
+//					return nil, err
+//				}
+//				return &MyEnv{
+//					VM: client.NewVM(vm),
+//				}, nil
+//			  },
+//		   })})
+//	  }
+//
+// Suite leverages pulumi features to compute the differences between the previous
+// environment and the new one to make environment updates faster.
+//
+// [testify Suite]: https://pkg.go.dev/github.com/stretchr/testify/suite
+type Suite[Env any] struct {
+	shortStackName string
+	fullStackName  string // From StackManager.Stack. Use shortStackName and the username
+	stackDef       *StackDefinition[Env]
+	destroyEnv     bool
+
+	// These fields are initialized in SetupTest
+	suite.Suite
+	Env              *Env
+	auth             client.Authentification
+	defaultConfigMap auto.ConfigMap
+}
+
+type StackDefinition[Env any] struct {
+	EnvCloudName string // Must be "aws/sandbox" for now.
+	EnvFactory   func(ctx *pulumi.Context) (*Env, error)
+	ConfigMap    auto.ConfigMap
+}
+
+// NewSuite creates a new Suite.
+// stackName is the name of the stack and should be unique.
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
Should be unique across suites, we should have a 1:1 relation between stack names and Suite ?

Reply:
> we should have a 1:1 relation between stack names and Suite ?
Most of the time yes but they may be some exceptions.